<?xml version="1.0" ?>
<robot name="backbone_manipulator">
    <material name="red">
        <color rgba="1 0 0 1"/>
    </material>

    <material name="green">
        <color rgba="0 1 0 1"/>
    </material>
    
    <material name="blue">
        <color rgba="0 0 1 1"/>
    </material>

    <material name="white">
        <color rgba="1 1 1 1"/>
    </material>

    <link name="base_link">
        <visual>
            <geometry>
                <cylinder length="0.01" radius="0.05"/>
            </geometry>
            <material name="white"/>
            <origin rpy="0 0 0" xyz="0 0 0.005"/>
        </visual>
        <collision>
            <geometry>
                <cylinder length="0.01" radius="0.05"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0.005"/>
        </collision>
    </link>

    <joint name="baselink_to_base" type="revolute">
        <parent link="base_link"/>
        <child link="base"/>
        <origin rpy="0 0 0" xyz="0 0 0.01"/>
        <axis xyz="0 0 1.0"/>
        <limit lower="-1.5708" upper="1.5708" velocity="1.0" effort="30" />
    </joint>

    <link name="base">
        <visual>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0.1"/>
            <material name="green"/>
        </visual>
        <collision>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0.1"/>
        </collision>
    </link>

    <joint name="base_to_network" type="revolute">
        <parent link="base"/>
        <child link="network_base_to_robot4"/>
        <origin rpy="0 0 0" xyz="0 0 0.1"/>
        <axis xyz="0 1.0 0"/>
        <limit lower="-1.5708" upper="1.5708" velocity="1.0" effort="30" />
    </joint>

    <link name="network_base_to_robot4">
        <visual>
            <geometry>
                <cylinder length="5" radius="0.005"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 2.5"/>
        </visual>
        <collision>
            <geometry>
                <cylinder length="5" radius="0.005"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 2.5"/>
        </collision>
    </link>

    <joint name="network_to_robot4" type="revolute">
        <parent link="network_base_to_robot4"/>
        <child link="robot4"/>
        <origin rpy="0 0 0" xyz="0 0 5"/>
        <axis xyz="0 0 1.0"/>
        <limit lower="-1.5708" upper="1.5708" velocity="1.0" effort="30" />
    </joint>

    <link name="robot4">
        <visual>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
        </collision>
    </link>

    <joint name="robot4_to_network" type="revolute">
        <parent link="robot4"/>
        <child link="network_robot4_to_robot3"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <axis xyz="0 1.0 0"/>
        <limit lower="-1.5708" upper="1.5708" velocity="1.0" effort="30" />
    </joint>

    <link name="network_robot4_to_robot3">
        <visual>
            <geometry>
                <cylinder length="5" radius="0.005"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 2.5"/>
        </visual>
        <collision>
            <geometry>
                <cylinder length="5" radius="0.005"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 2.5"/>
        </collision>
    </link>

    <joint name="network_to_robot3" type="revolute">
        <parent link="network_robot4_to_robot3"/>
        <child link="robot3"/>
        <origin rpy="0 0 0" xyz="0 0 5"/>
        <axis xyz="0 0 1.0"/>
        <limit lower="-1.5708" upper="1.5708" velocity="1.0" effort="30" />
    </joint>

    <link name="robot3">
        <visual>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
        </collision>
    </link>

   <joint name="robot3_to_network" type="revolute">
        <parent link="robot3"/>
        <child link="network_robot3_to_robot2"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <axis xyz="0 1.0 0"/>
        <limit lower="-1.5708" upper="1.5708" velocity="1.0" effort="30" />
    </joint>

    <link name="network_robot3_to_robot2">
        <visual>
            <geometry>
                <cylinder length="5" radius="0.005"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 2.5"/>
        </visual>
        <collision>
            <geometry>
                <cylinder length="5" radius="0.005"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 2.5"/>
        </collision>
    </link>

    <joint name="network_to_robot2" type="revolute">
        <parent link="network_robot3_to_robot2"/>
        <child link="robot2"/>
        <origin rpy="0 0 0" xyz="0 0 5"/>
        <axis xyz="0 0 1.0"/>
        <limit lower="-1.5708" upper="1.5708" velocity="1.0" effort="30" />
    </joint>

    <link name="robot2">
        <visual>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
        </collision>
    </link>

    <joint name="robot2_to_network" type="revolute">
        <parent link="robot2"/>
        <child link="network_robot2_to_robot1"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <axis xyz="0 1.0 0"/>
        <limit lower="-1.5708" upper="1.5708" velocity="1.0" effort="30" />
    </joint>

    <link name="network_robot2_to_robot1">
        <visual>
            <geometry>
                <cylinder length="5" radius="0.005"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 2.5"/>
        </visual>
        <collision>
            <geometry>
                <cylinder length="5" radius="0.005"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 2.5"/>
        </collision>
    </link>

    <joint name="network_to_robot1" type="revolute">
        <parent link="network_robot2_to_robot1"/>
        <child link="robot1"/>
        <origin rpy="0 0 0" xyz="0 0 5"/>
        <axis xyz="0 0 1.0"/>
        <limit lower="-1.5708" upper="1.5708" velocity="1.0" effort="30" />
    </joint>

    <link name="robot1">
        <visual>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
        </collision>
    </link>

   <joint name="robot1_to_network" type="revolute">
        <parent link="robot1"/>
        <child link="network_robot1_to_robot0"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <axis xyz="0 1.0 0"/>
        <limit lower="-1.5708" upper="1.5708" velocity="1.0" effort="30" />
    </joint>

    <link name="network_robot1_to_robot0">
        <visual>
            <geometry>
                <cylinder length="5" radius="0.005"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 2.5"/>
        </visual>
        <collision>
            <geometry>
                <cylinder length="5" radius="0.005"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 2.5"/>
        </collision>
    </link>

    <joint name="network_to_robot0" type="fixed">
        <parent link="network_robot1_to_robot0"/>
        <child link="robot0"/>
        <origin rpy="0 0 0" xyz="0 0 5"/>
    </joint>

    <link name="robot0">
        <visual>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <sphere radius="0.1"/>
            </geometry>
            <origin rpy="0 0 0" xyz="0 0 0"/>
        </collision>
    </link>
</robot>